### PR TITLE
WhenExpressions in Finally Tasks

### PR DESCRIPTION
**Pull request by jerop**
# Changes

Users can guard execution of `Tasks` using `WhenExpressions`, but that
is currently not supported in `Finally Tasks`.

This change adds support for `WhenExpressions` in `Finally Tasks` not
only to provide efficient guarded execution but also to improve the
reusability of `Tasks` in `Finally`. The proposal is described further in
the [WhenExpressions in Finally Tasks TEP](https://github.com/tektoncd/community/blob/master/teps/0045-whenexpressions-in-finally-tasks.md).

Given we've recently added support for `Results` and `Status` in
`Finally Tasks`, this is an opportune time to enable `WhenExpressions`
in `Finally Tasks`.

/kind feature

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
WhenExpressions are supported in Finally Tasks not only to provide efficient guarded execution but also to improve the
reusability of `Tasks` in `Finally`
```

### CODE REVIEW

**Comment by jerop** [comment]
closes https://github.com/tektoncd/pipeline/issues/3438

**Comment by pritidesai** [comment]
thanks @jerop, I haven't reviewed the PR yet but I would like to see `e2e` tests in [pipelinefinally_test.go](https://github.com/tektoncd/pipeline/blob/master/test/pipelinefinally_test.go)

**Comment by jerop** [comment]
> thanks @jerop, I haven't reviewed the PR yet but I would like to see `e2e` tests in [pipelinefinally_test.go](https://github.com/tektoncd/pipeline/blob/master/test/pipelinefinally_test.go)

i'd added them in [pipelinerun_test.go](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/pipelinerun/pipelinerun_test.go), will add in [pipelinefinally_test.go](https://github.com/tektoncd/pipeline/blob/master/test/pipelinefinally_test.go) as well, thanks @pritidesai! 

/hold

**Comment by jerop** [comment]
@pritidesai I've added the tests in [pipelinefinally_test.go](https://github.com/tektoncd/pipeline/blob/master/test/pipelinefinally_test.go), please take a look 🙏🏾 

/hold cancel

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 1]
```diff
@@ -4855,10 +4875,12 @@ func TestReconcileWithTaskResultsInFinalTasks(t *testing.T) {
 	}
 	expectedSkippedTasks := []v1beta1.SkippedTask{{
 		Name: "final-task-2",
+	}, {
+		Name: "final-task-3",
 	}}
 
 	if d := cmp.Diff(expectedSkippedTasks, reconciledRun.Status.SkippedTasks); d != "" {
-		t.Fatalf("Expected to see only one final task (final-task-2) in the list of skipped tasks. Diff: %s", diff.PrintWantGot(d))
+		t.Fatalf("Didn't get the expected the list of skipped tasks. Diff: %s", diff.PrintWantGot(d))
```
Comment (t-persson) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:4883]:
```suggestion
		t.Fatalf("Didn't get the expected list of skipped tasks. Diff: %s", diff.PrintWantGot(d))
```

File: docs/pipelines.md [hunk 2]
```diff
@@ -880,10 +900,119 @@ This kind of variable can have any one of the values from the following table:
...
+      description: a boolean indicating whether the notifications should be sent
+    tasks:
+      - name: golang-build
+        taskRef:
+          name: golang-build
+      # […]
+    finally:
+      - name: notify-build-failure # executed only when build task fails and notifications are enabled
+        when:
+          - input: $(tasks.golang-build.status)
```
Comment (chmouel) [inline docs/pipelines.md:935]: I see most of the examples or tests refer to task status, is there a way to guard again pipeline run status? I think the user in the send to slack example would typically would want to grab the status of the execution of the whole pipeline if it has failed or succeeded, what do you think having this as the example?

Comment (jerop) [reply docs/pipelines.md:935]: pipeline run status is calculated based on both tasks and finally tasks, as described [here](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#pipelinerun-status-with-finally), so I don't see how we can guard against that because it won't be available yet when executing finally tasks

if you're referring to the overall status of the tasks only, not finally tasks, I can imagine that a user could pass in all the statuses of the tasks to the values array, as shown below 

```yaml
finally:
      - name: notify-failure # executed only when tasks failed
        when:
          - input: "Failed"
 	    operator: in
            values: [$(tasks.unit-tests.status), $(tasks.golang-build.status), $(tasks.deploy.status)]         
```

Comment (jerop) [reply docs/pipelines.md:935]: /cc @pritidesai on execution status variables available in finally

**Review by sbwsg — APPROVED** [review]

File: docs/pipelines.md (hunk 2 above)
Comment (pritidesai) [reply docs/pipelines.md:935]: @chmouel I have a WIP proposal to support [aggregate](https://github.com/tektoncd/pipeline/issues/1020#issuecomment-747156100) status of `tasks` section, would that work for your use case? 

The proposal includes introducing a new context variable `$(tasks.status)` (feedback on syntax is welcome) 🙏. `$(tasks.status)` would be set to `Failed` if any on of the `dag` task fails, and `Succeeded` if all `dag` tasks succeed.

Comment (jerop) [reply docs/pipelines.md:935]: the aggregate status of `tasks` would be a very useful in `whenexpressions` in `finally tasks`, thanks for the update @pritidesai!

File: docs/pipelines.md [hunk 3]
```diff
@@ -6,27 +6,47 @@ weight: 3
 -->
 # Pipelines
 
-- [Overview](#pipelines)
-- [Configuring a `Pipeline`](#configuring-a-pipeline)
+- [Pipelines](#pipelines)
```
Comment (bobcatfish) [inline docs/pipelines.md:9]: i think something odd happened here - not sure why we have a top level link back to the top of the page? (maybe you're using a tool that's doing that?)

maybe these updates to the TOC would make sense in a separate PR?  i think i like having the more verbose TOC but it will be harder to keep up to date 🤔

File: docs/pipelines.md [hunk 4]
```diff
@@ -880,10 +900,119 @@ This kind of variable can have any one of the values from the following table:
...
+          - input: $(params.enable-notifications)
+            operator: in
+            values: ["true"]
+        taskRef:
+          name: send-to-slack-channel
+      # […]
+  params:
+    - name: enable-notifications
+      value: true
+```
```
Comment (bobcatfish) [inline docs/pipelines.md:947]: cool example :D

File: docs/pipelines.md [hunk 5]
```diff
@@ -880,10 +900,119 @@ This kind of variable can have any one of the values from the following table:
...
+        taskRef:
+          name: boskos-acquire
+      - name: use-resource
+      # […]
+    finally:
+      - name: boskos-release # executed only when leased resource is phonetic-project
+        when:
+          - input: $(tasks.boskos-acquire.results.leased-resource)
+            operator: in
+            values: ["phonetic-project"]
```
Comment (bobcatfish) [inline docs/pipelines.md:972]: this example feels a bit weird - not sure why we are releasing only in this one case, feels more likely that this would be checking the status of the boskos-acquire task 🤔   np if you want to leave it - alternative idea, git-clone task maybe?

alternatively: maybe we don't need to have an example of results + finally? knowing how to use variable replacement in general could be enough?

File: docs/pipelines.md [hunk 6]
```diff
@@ -880,10 +900,119 @@ This kind of variable can have any one of the values from the following table:
...
+        taskRef:
+          name: boskos-release
+      # […]
+```
+
+If the `WhenExpressions` in a `Finally Task` use `Results` from a skipped or failed non-finally `Tasks`, then the
+`Finally Task` would also be skipped and be included in the list of `Skipped Tasks` in the `Status`, [similarly to when using
+`Results` in other parts of the `Finally Task`](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#consuming-task-execution-results-in-finally).
```
Comment (bobcatfish) [inline docs/pipelines.md:980]: you can also link to this with `pipeline.md#consuming-task...` (i..e you dont need the whole URL, and using the short form means github can preview the link in PRs etc)

File: docs/pipelines.md [hunk 7]
```diff
@@ -880,10 +900,119 @@ This kind of variable can have any one of the values from the following table:
...
+          name: golang-build
+      # […]
+    finally:
+      - name: notify-build-failure # executed only when build task fails
+        when:
+          - input: $(tasks.golang-build.status)
+            operator: in
+            values: ["Failed"]
+        taskRef:
+          name: send-to-slack-channel
```
Comment (bobcatfish) [inline docs/pipelines.md:1007]: woooot

File: docs/pipelines.md (hunk 3 above)
Comment (bobcatfish) [reply docs/pipelines.md:9]: (feel free to ignore me)

File: test/pipelinefinally_test.go [hunk 8]
```diff
@@ -84,6 +85,11 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_InvalidTaskResult_Failure(t *test
 		t.Fatalf("Failed to create Task producing task results: %s", err)
 	}
 
+	taskConsumingResultInParam := getSuccessTaskConsumingResultsInParam(t, namespace)
+	if _, err := c.TaskClient.Create(ctx, taskConsumingResultInParam, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to create Task consuming task results in param: %s", err)
+	}
+
 	taskConsumingResult := getSuccessTaskConsumingResults(t, namespace)
```
Comment (bobcatfish) [inline test/pipelinefinally_test.go:93]: im finding this a bit hard to follow, i think mostly b/c the details of the tasks are hidden in the functions (i.e. it was confusing before this change!), im wondering if some comments could help explain, e.g. what is taskConsumingResultInParam and how is it different from taskConsumingResult and why do we need both

File: test/pipelinefinally_test.go [hunk 9]
```diff
@@ -201,13 +247,16 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_InvalidTaskResult_Failure(t *test
 		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", pipelineRun.Name, err)
 	}
 
-	// expecting taskRuns for dagtask1, dagtask2, dagtask3 (with condition failure), dagtask5, finaltask1, finaltask2, and finaltaskconsumingdagtask5
-	if len(taskrunList.Items) != 7 {
+	// expecting taskRuns for dagtask1, dagtask2, dagtask3 (with condition failure), dagtask5, finaltask1, finaltask2,
+	// finaltaskconsumingdagtask5, finaltaskconsumingdagtask5inwhenexpressionstrue, finaltaskconsumingdagtask5executionstatusinwhenexpressionstrue
```
Comment (bobcatfish) [inline test/pipelinefinally_test.go:251]: i feel like these names are getting a bit long - i wonder if there is some other way we can make this clearer 🤔

File: test/pipelinefinally_test.go [hunk 10]
```diff
@@ -201,13 +247,16 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_InvalidTaskResult_Failure(t *test
...
+	expectedTaskRunsCount := 9
+	if len(taskrunList.Items) != expectedTaskRunsCount {
 		var s []string
 		for _, n := range taskrunList.Items {
 			s = append(s, n.Labels["tekton.dev/pipelineTask"])
 		}
-		t.Fatalf("Error retrieving TaskRuns for PipelineRun %s. Expected 5 taskRuns and found taskRuns for: %s", pipelineRun.Name, strings.Join(s, ", "))
+		t.Fatalf("Error retrieving TaskRuns for PipelineRun %s. Expected %s taskRuns and found %s taskRuns for: %s",
+			pipelineRun.Name, strconv.Itoa(expectedTaskRunsCount), strconv.Itoa(len(taskrunList.Items)), strings.Join(s, ", "))
```
Comment (bobcatfish) [inline test/pipelinefinally_test.go:259]: instead of calling ` strconv.Itoa` we could use the `%d` format operator (%s is forcing us to use a string: https://golang.org/pkg/fmt/#hdr-Printing)

File: test/pipelinefinally_test.go [hunk 11]
```diff
@@ -269,7 +318,17 @@ func TestPipelineLevelFinally_OneDAGTaskFailed_InvalidTaskResult_Failure(t *test
 					t.Errorf("Error resolving task result reference in a finally task %s", n)
 				}
 			}
-		case n == "finaltaskconsumingdagtask1" || n == "finaltaskconsumingdagtask4":
+		case n == "finaltaskconsumingdagtask5inwhenexpressionstrue":
+			if err := WaitForTaskRunState(ctx, c, taskrunItem.Name, TaskRunSucceed(taskrunItem.Name), "TaskRunSuccess"); err != nil {
```
Comment (bobcatfish) [inline test/pipelinefinally_test.go:322]: do we want to be waiting here? I think we already waited for the pipelinerun as a whole to get into the state we expected (or maybe im wrong?) - if the taskruns aren't in the expected state after the pipelinerun has completed, i think something has gone wrong?

File: test/pipelinefinally_test.go [hunk 12]
```diff
@@ -477,12 +556,17 @@ func getPipeline(t *testing.T, namespace string, dag map[string]pipelineTask, f
...
-			Params:  v.Param,
-		})
+		}
+		if len(v.Param) != 0 {
+			finalTask.Params = v.Param
+		}
+		if len(v.When) != 0 {
+			finalTask.WhenExpressions = v.When
+		}
+		fpt = append(fpt, finalTask)
```
Comment (bobcatfish) [inline test/pipelinefinally_test.go:569]: i wonder how useful this function actually is vs declaring a pipeline inline

**Review by bobcatfish — COMMENTED** [review]
So far I've reviewed docs, examples and the end to end test - planning to come back and review the rest but dont want to lose this so far :)

File: docs/pipelines.md (hunk 3 above)
Comment (jerop) [reply docs/pipelines.md:9]: moved the TOC updates to a separate PR: https://github.com/tektoncd/pipeline/pull/3743

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 13]
```diff
@@ -4855,10 +4875,12 @@ func TestReconcileWithTaskResultsInFinalTasks(t *testing.T) {
...
 		Name: "final-task-2",
+	}, {
+		Name: "final-task-3",
 	}}
 
 	if d := cmp.Diff(expectedSkippedTasks, reconciledRun.Status.SkippedTasks); d != "" {
-		t.Fatalf("Expected to see only one final task (final-task-2) in the list of skipped tasks. Diff: %s", diff.PrintWantGot(d))
+		t.Fatalf("Didn't get the expected list of skipped tasks. Diff: %s", diff.PrintWantGot(d))
 	}
 }
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:4885]: nice update!

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go [hunk 14]
```diff
@@ -361,10 +404,15 @@ func TestPipelineSpec_Validate_Failure(t *testing.T) {
...
+			Finally: []PipelineTask{{
+				Name:            "invalid-pipeline-task-finally",
+				TaskRef:         &TaskRef{Name: "foo-task"},
+				WhenExpressions: []WhenExpression{{}},
+			}},
 		},
 		expectedError: apis.FieldError{
 			Message: `missing field(s)`,
-			Paths:   []string{"tasks[1].when[0]"},
+			Paths:   []string{"tasks[1].when[0]", "finally[0].when[0]"},
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:415]: im a little nervous (which maybe is a sign that im overthinking this!) about all the test here combining the two scenarios, i.e. validation failure in the pipelinetask AND in the finally task.

at least since the path contains both we can know both were caught, but im wondering if it'd be good to also cover a case where only one is set.

since the new cases we want to add are so similar to the existing ones, we have a few choices:

1. Combine the cases together like this PR does
2. Duplicate the cases
3. To be extra sure, do (2) AND do (1) (test them all separately AND together)

I think (3) is overkill, so I'm happy to go with this - I feel like if we do (1) we should do (2) for _at least_ one scenario, or do (2) and have one test that combines the cases like in (1)

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go [hunk 15]
```diff
@@ -486,11 +493,14 @@ func validateTasksInputFrom(tasks []PipelineTask) (errs *apis.FieldError) {
 	return errs
 }
 
-func validateWhenExpressions(tasks []PipelineTask) (errs *apis.FieldError) {
+func validateWhenExpressions(tasks []PipelineTask, finalTasks []PipelineTask) (errs *apis.FieldError) {
 	for i, t := range tasks {
 		errs = errs.Also(validateOneOfWhenExpressionsOrConditions(t).ViaFieldIndex("tasks", i))
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1beta1/pipeline_validation.go:498]: maybe this is overkill, but now that we're allowing when expressions in finally, does it make sense to add a test that ensures that conditions aren't allows in when expressions? (just ignore me if this already exists)

File: pkg/reconciler/pipelinerun/resources/apply_test.go [hunk 16]
```diff
@@ -1062,6 +1082,11 @@ func TestApplyTaskRunContext(t *testing.T) {
 				Name:  "task3",
 				Value: *v1beta1.NewArrayOrString("$(tasks.task3.status)"),
 			}},
+			WhenExpressions: v1beta1.WhenExpressions{{
+				Input:    "$(tasks.task1.status)",
+				Operator: selection.In,
+				Values:   []string{"$(tasks.task3.status)"},
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/apply_test.go:1088]: can we add coverage for results replacements in when expressions via the tests in this file as well? (i only see status and param here, and only see results replacement tested in reconciler, end to end and examples)

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 17]
```diff
@@ -93,6 +93,7 @@ func ApplyPipelineTaskContext(state PipelineRunState, replacements map[string]st
 		if resolvedPipelineRunTask.PipelineTask != nil {
 			pipelineTask := resolvedPipelineRunTask.PipelineTask.DeepCopy()
 			pipelineTask.Params = replaceParamValues(pipelineTask.Params, replacements, nil)
+			pipelineTask.WhenExpressions = pipelineTask.WhenExpressions.ReplaceWhenExpressionsVariables(replacements)
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/apply.go:96]: I'm a bit confused by this line - this doesn't seem specific to finally tasks, im wondering why pipelinetasks with when expressions currently work without this line? (was there a bug? if so, might make sense to separate that out into a separate commit and/or PR if possible)

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 18]
```diff
@@ -156,12 +156,20 @@ func (t *ResolvedPipelineRunTask) checkParentsDone(facts *PipelineRunFacts) bool
...
+	if facts.isFinalTask(t.PipelineTask.Name) {
+		if t.whenExpressionsSkip(facts) {
+			return true
+		}
+	}
+
+	if facts.isDAGTask(t.PipelineTask.Name) {
+		if t.conditionsSkip() || t.whenExpressionsSkip(facts) || t.parentTasksSkip(facts) || facts.IsStopping() {
+			return true
+		}
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:172]: nit: is it possible to word these as a switch statement? feels weird that facts.isFinalTask can be true and yet facts.isDAGTask(t.PipelineTask.Name) still gets evaluated

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go [hunk 19]
```diff
@@ -972,32 +1007,38 @@ func TestIsSkipped(t *testing.T) {
...
 		expected: map[string]bool{
 			"mytask13": false,
 		},
 	}, {
 		name:  "run-cancelled",
-		state: runCancelled,
+		tasks: runCancelled,
 		expected: map[string]bool{
 			"mytask13": false,
 		},
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1019]: similar to above comments about combining cases together, i wonder if it would make sense to add a few new ones here too?

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go [hunk 20]
```diff
@@ -972,32 +1007,38 @@ func TestIsSkipped(t *testing.T) {
...
 	}} {
 		t.Run(tc.name, func(t *testing.T) {
-			d, err := dagFromState(tc.state)
+			d, err := dagFromState(tc.tasks)
+			if err != nil {
+				t.Fatalf("Could not get a dag from the TC state %#v: %v", tc.tasks, err)
+			}
+			f, err := dagFromState(tc.finally)
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1026]: there's something very odd about getting a "dag" from the finally tasks - but i think thats a whole other story XD

**Review by bobcatfish — COMMENTED** [review]
Overall looks great!!! All of my comments are pretty much nitpicks or just wondering if we want (EVEN MORE) unit tests

Very exciting to see this coming together of the features you and @pritidesai have been working on (finally, when expressions, status)!!!! 💃

File: pkg/apis/pipeline/v1beta1/pipeline_validation.go (hunk 15 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation.go:498]: > that conditions aren't allows in when expressions 

did you mean "that conditions aren't allowed in finally"? if so, that's already tested in:

https://github.com/tektoncd/pipeline/blob/075910e6a0e7e40fb01530aab05d4ec8ac5433a7/pkg/apis/pipeline/v1beta1/pipeline_validation_test.go#L2001-L2014

File: docs/pipelines.md (hunk 7 above)
Comment (jerop) [reply docs/pipelines.md:994]: 🎉

File: pkg/reconciler/pipelinerun/resources/apply.go (hunk 17 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/apply.go:96]: The `ApplyPipelineTaskContext` function replaces the context variables for execution status. It is written more generally, I guess so it could be easily reused later, but it is currently used on finally tasks only in:

https://github.com/tektoncd/pipeline/blob/08177fab0e2f2d2cc8e30e67f018a552338fca26/pkg/reconciler/pipelinerun/pipelinerun.go#L614-L617

so the line added in this PR is to substitute execution status from pipeline tasks to the when expressions in the finally tasks, so there's no bug

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 18 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:170]: simplified the logic and removed the double checks

File: test/pipelinefinally_test.go (hunk 9 above)
Comment (jerop) [reply test/pipelinefinally_test.go]: shortened them and added comments to describe them

File: test/pipelinefinally_test.go (hunk 10 above)
Comment (jerop) [reply test/pipelinefinally_test.go]: replaced it, thanks!

File: test/pipelinefinally_test.go (hunk 11 above)
Comment (jerop) [reply test/pipelinefinally_test.go]: that's right, replaced with checking the taskrun status

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 14 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:523]: we can't have `finally tasks` without `dag tasks`, so we can't do option 2 with `finally tasks` only

left the `dag tasks` as they are, and added cases with combined `dag tasks` and `finally tasks`

File: pkg/reconciler/pipelinerun/resources/apply_test.go (hunk 16 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/apply_test.go:1088]: result replacements in when expressions are already tested in this file at: https://github.com/tektoncd/pipeline/blob/33c3a6972691924aa011d15bc7307b5a19775bcb/pkg/reconciler/pipelinerun/resources/apply_test.go#L291-L447

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go (hunk 19 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1019]: will separate them, waiting for this to get in https://github.com/tektoncd/pipeline/pull/3744

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go (hunk 20 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1026]: created a PR for the refactor so that we consider them separately: https://github.com/tektoncd/pipeline/pull/3744

**Comment by jerop** [comment]
moved out the testing refactors to a separate PR that we need to get merged first: https://github.com/tektoncd/pipeline/pull/3744

/hold

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go (hunk 19 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1045]: added new ones instead of combining

File: test/pipelinefinally_test.go (hunk 12 above)
Comment (jerop) [reply test/pipelinefinally_test.go:573]: @pritidesai what do you think? maybe we can refactor this in a future PR?

File: test/pipelinefinally_test.go (hunk 8 above)
Comment (jerop) [reply test/pipelinefinally_test.go]: refactored to reuse the same function, which tasks the params list as an argument

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 18 above)
Comment (pritidesai) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:170]: This could be simplified by utilizing `IsFinallySkipped` instead of changing this function.

https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L220

This skip function is primarily used to build cache of skipped `dag` tasks. This caching was introduced in PR -https://github.com/tektoncd/pipeline/pull/3524

https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L59

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 21]
```diff
@@ -193,7 +199,7 @@ func (t *ResolvedPipelineRunTask) conditionsSkip() bool {
 }
 
 func (t *ResolvedPipelineRunTask) whenExpressionsSkip(facts *PipelineRunFacts) bool {
-	if t.checkParentsDone(facts) {
+	if facts.isFinalTask(t.PipelineTask.Name) || t.checkParentsDone(facts) {
```
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:202]: this change is not needed if we are utilizing `IsFinallySkipped`

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 18 above)
Comment (pritidesai) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:170]: Just introduce one `if` statement after [line 227](https://github.com/tektoncd/pipeline/blob/01286cf978750e85609fa0a0e9a1aa177fc724f8/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L227) 

```
		if t.whenExpressionsSkip(facts) {
			return true
		}

```

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go [hunk 22]
```diff
@@ -939,9 +940,39 @@ func TestIsSkipped(t *testing.T) {
 		expected: map[string]bool{
 			"mytask10": false,
 		},
+	}, {
+		name: "tasks-when-expressions-in-finally-passed",
```
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:944]: these tests can be moved to `TestResolvedPipelineRunTask_IsFinallySkipped`

Comment (pritidesai) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:944]: PR - https://github.com/tektoncd/pipeline/pull/3744 will not be needed if we change `IsFinallySkipped`.

**Comment by jerop** [comment]
thanks for the review @pritidesai 🙏🏾 

changed to using `IsFinallySkipped` and closed https://github.com/tektoncd/pipeline/pull/3744

/hold cancel

File: pkg/apis/pipeline/v1beta1/pipeline_validation_test.go (hunk 14 above)
Comment (jerop) [reply pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:655]: actually ended up doing (3), but the ones with finally tasks have dag tasks as well (but the dag tasks in some of those cases don't have when expressions)

File: pkg/reconciler/pipelinerun/resources/apply.go (hunk 17 above)
Comment (bobcatfish) [reply pkg/reconciler/pipelinerun/resources/apply.go:96]: ahh i see: since it was only used for finally tasks (and since we dont support status in non finally tasks), got it! thanks for explaining :D

**Comment by pritidesai** [comment]
@jerop `GetSkippedTasks` adds the finally tasks into the list of skippedTasks:

https://github.com/tektoncd/pipeline/blob/98507859fa961f2152b65ad911defc2f0a2ef804/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L369-L386

Can we please set WhenExpressions for finally tasks when they are specified?

```
		if rprt.IsFinallySkipped(facts) {
			skippedTask := v1beta1.SkippedTask{
				Name: rprt.PipelineTask.Name,
			}
			if rprt.PipelineTask.WhenExpressions != nil || len(rprt.PipelineTask.WhenExpressions) != 0 {
				skippedTask.WhenExpressions = rprt.PipelineTask.WhenExpressions
			}
			skipped = append(skipped, skippedTask)
		}
```

such that finally tasks being in the list of skipped tasks have the when expressions initialized:

```
    {
        "name": "finally-task-should-be-skipped-1"
      },
      {
        "name": "finally-task-should-be-skipped-2"
      },
      {
        "name": "finally-task-should-be-skipped-3"
      }
```

**Comment by jerop** [comment]
thanks for the review @pritidesai! 🙏🏾 

> Can we please set `WhenExpressions` for finally tasks when one or more expressions are specified?
> 
> ```
> 		if rprt.IsFinallySkipped(facts) {
> 			skippedTask := v1beta1.SkippedTask{
> 				Name: rprt.PipelineTask.Name,
> 			}
> 			if rprt.PipelineTask.WhenExpressions != nil || len(rprt.PipelineTask.WhenExpressions) != 0 {
> 				skippedTask.WhenExpressions = rprt.PipelineTask.WhenExpressions
> 			}
> 			skipped = append(skipped, skippedTask)
> 		}
> ```

makes sense to do it similarly to `Tasks`, but did it a bit differently to ensure we don't include `WhenExpressions` in `Finally Tasks` that were skipped because they referenced missing `Results` (not because their `WhenExpressions` evaluated to `False`):

https://github.com/tektoncd/pipeline/blob/c719b6a12fec2e4d0d5d147b451d7fc991b9638b/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L379-L389

didn't check for nil because it doesn't change anything, since it defaults to nil anyway, like in the `Tasks` part of the function:
https://github.com/tektoncd/pipeline/blob/c719b6a12fec2e4d0d5d147b451d7fc991b9638b/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L372-L378

**Comment by pritidesai** [comment]
thanks a bunch @jerop, great point 🙇‍♀️

File: pkg/reconciler/pipelinerun/pipelinerun_test.go [hunk 23]
```diff
@@ -4714,6 +4714,71 @@ func TestReconcileWithTaskResultsInFinalTasks(t *testing.T) {
...
+						StringVal: "param",
+					}},
+				},
+				WhenExpressions: v1beta1.WhenExpressions{{
+					Input:    "$(tasks.dag-task-1.results.aResult)",
+					Operator: selection.NotIn,
+					Values:   []string{"aResultValue"},
+				}},
+			}, {
+				// final task executed because when expressions evaluated to false
```
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:4734]: NIT: s/executed/skipped/

File: test/pipelinefinally_test.go (hunk 12 above)
Comment (pritidesai) [reply test/pipelinefinally_test.go:573]: sure we can in a separate PR. This function is useful since its shared between two separate tests. We had the tests implemented without this common function and found a lot of repetition. We can declare a pipeline inline in favor of the readability instead of using this function.

**Comment by pritidesai** [comment]
Thanks a bunch @jerop for this feature.

I am having hard time understanding validation unit tests. I feel we have many overlapping tests. Ideally, would like to see the separate unit tests validating when expression syntax e.g. input and values are initialized, a valid operator is specified, etc. Now once we have validated the syntax, I don't see a need to have separate tests to validate the syntax again as part of either `tasks` or `finally`. Rather validate `when` expressions references (e.g. referencing a task result or defined parameter) with the `tasks` and `finally` sections. This can be refactored and simplified later and will not block this PR.

/lgtm

File: pkg/reconciler/pipelinerun/pipelinerun_test.go (hunk 23 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go]: good catch, fixed it

**Comment by jerop** [comment]
@pritidesai I agree, we need to refactor the validation testing -- can open an issue to track that

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 24]
```diff
@@ -630,7 +630,7 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	}
 
 	for _, rprt := range nextRprts {
-		if rprt == nil || rprt.Skip(pipelineRunFacts) {
+		if rprt == nil || rprt.Skip(pipelineRunFacts) || rprt.IsFinallySkipped(pipelineRunFacts) {
```
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/pipelinerun.go:633]: hey @jerop, based on my understanding, `rprt.Skip(pipelineRunFacts)` returns `false` for any finally task but `rprt.IsFinallySkipped(pipelineRunFacts)` evaluates when expression and returns `true` since the when expression evaluates to `false`.

Comment (pritidesai) [reply pkg/reconciler/pipelinerun/pipelinerun.go:633]: but this would work, provided `nextRprts` contains the finally tasks 🤣
